src/System.Private.CoreLib/shared/System/Text/UTF8Encoding.cs

```diff
@@ -377,2120 +539,245 @@ public override unsafe int GetCharCount(ReadOnlySpan<byte> bytes)
         public unsafe override int GetChars(byte* bytes, int byteCount, char* chars, int charCount)
         {
             // Validate Parameters
-            if (bytes == null || chars == null)
-                throw new ArgumentNullException(bytes == null ? nameof(bytes) : nameof(chars), SR.ArgumentNull_Array);
-
-            if (charCount < 0 || byteCount < 0)
-                throw new ArgumentOutOfRangeException((charCount < 0 ? nameof(charCount) : nameof(byteCount)), SR.ArgumentOutOfRange_NeedNonNegNum);
 
-            return GetChars(bytes, byteCount, chars, charCount, null);
-        }
+            if (bytes is null || chars is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(
+                    argument: (bytes is null) ? ExceptionArgument.bytes : ExceptionArgument.chars,
+                    resource: ExceptionResource.ArgumentNull_Array);
+            }
 
-        public override unsafe int GetChars(ReadOnlySpan<byte> bytes, Span<char> chars)
-        {
-            fixed (byte* bytesPtr = &MemoryMarshal.GetNonNullPinnableReference(bytes))
-            fixed (char* charsPtr = &MemoryMarshal.GetNonNullPinnableReference(chars))
+            if ((byteCount | charCount) < 0)
             {
-                return GetChars(bytesPtr, bytes.Length, charsPtr, chars.Length, baseDecoder: null);
+                ThrowHelper.ThrowArgumentOutOfRangeException(
+                    argument: (byteCount < 0) ? ExceptionArgument.byteCount : ExceptionArgument.charCount,
+                    resource: ExceptionResource.ArgumentOutOfRange_NeedNonNegNum);
```

Comment:
Feels like you do this enough that you'd have a dedicated demuxer helper.

```C#
ThrowHelper.ThrowNeedNonNegNum(
    byteCount,
    ExceptionArgument.byteCount,
    ExceptionArgument.charCount);

...

void ThrowNeedNonNegNum(int firstVal, string firstName, string secondName)
{
    ThrowArgumentOutOfRangeException(
        firstVal < 0 ? firstName : secondName,
        ExceptionResource.ArgumentOutOfRange_NeedNonNegNum);
}
```

Just to cut down on the size of the boilerplate.

Reply:
Agree. Amusingly enough, it's written this way because it minimized the diff in the _ASCIIEncoding.cs__ case, and the code got copied to this file so that the diff between __ASCIIEncoding.cs__ and __UTF8Encoding.cs__ would be minimized.

There are some other changes / optimizations I want to make here that should further reduce the overhead of these method calls. Will open a new PR so that I can address them in __ASCIIEncoding.cs__ and here at the same time.